assets/js/RepLog/RepLogApp.js

```diff
@@ -0,0 +1,140 @@
+import React, { Component } from 'react';
+import RepLogs from './RepLogs';
+import PropTypes from 'prop-types';
+import uuid from 'uuid/v4';
+import { getRepLogs, deleteRepLog, createRepLog } from '../api/rep_log_api';
+
+export default class RepLogApp extends Component {
+    constructor(props) {
+        super(props);
+
+        this.state = Object.assign({
+            highlightedRowId: null,
+            repLogs: [],
+            numberOfHearts: 1,
+            isLoaded: false,
+            isSavingNewRepLog: false,
+            successMessage: '',
+            newRepLogValidationErrorMessage: '',
+            itemOptions: []
+        }, this.props.initialState);
+
+        this.handleRowClick = this.handleRowClick.bind(this);
+        this.handleAddRepLog = this.handleAddRepLog.bind(this);
+        this.handleHeartChange = this.handleHeartChange.bind(this);
+        this.handleDeleteRepLog = this.handleDeleteRepLog.bind(this);
+    }
+
+    componentDidMount() {
+        getRepLogs()
+            .then((data) => {
+                this.setState({
+                    repLogs: data,
+                    isLoaded: true
+                })
+            });
+    }
+
+    componentWillUnmount() {
+        if (this.clearSuccessMessageTimeout) {
+            clearTimeout(this.clearSuccessMessageTimeout);
+        }
+    }
+
+    handleRowClick(repLogId) {
+        this.setState({highlightedRowId: repLogId});
+    }
+
+    handleAddRepLog(item, quantity) {
+        const newRep = {
+            reps: quantity,
+            item: item
+        };
+
+        this.setState({
+            isSavingNewRepLog: true
+        });
+
+        createRepLog(newRep)
+            .then(repLog => {
+                this.setState(prevState => {
+                    const newRepLogs = [...prevState.repLogs, repLog];
+
+                    return {
+                        repLogs: newRepLogs,
+                        newRepLogValidationErrorMessage: '',
+                    };
+                });
+
+                this.setSuccessMessage('Rep Log Saved!');
+            })
+            .catch(error => {
+                error.response.json().then(errorsData => {
+                    const errors = errorsData.errors;
+                    const firstError = errors[Object.keys(errors)[0]];
+
+                    this.setState({
+                        newRepLogValidationErrorMessage: firstError
+                    });
+                })
+            })
+            .finally(() => {
```

Comment:
The finally case will cause a double render. It's better to merge this in with the other setState calls. You can merge one or more properties in by defining them first and then use the `...` notation when passing the object to `setState`:

```jsx
const defaults = { isSavingNewRepLog: false };
this.setState({ ...defaults, other: 'options' });
```

This requires this babel plugin, but it's so much more expressive: https://babeljs.io/docs/plugins/transform-object-rest-spread/